File: segundoPeriodo/AEDS-2/estudos/TADsLineares/pilha/c/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *array;
    int topo;
    int tam;
} Pilha;

void criarPilha(Pilha *p, int x) {
    p->array = (int*)malloc(x * sizeof(int));
    p->topo = -1;
    p->tam = x;
}

void inserir(Pilha *p, int x) {
    if(p->tam - 1 == p->topo) {
        printf("Pilha cheia\n");
    }
    p->array[++p->topo] = x;
}

int remover(Pilha *p) {
    if(p->topo == -1) {
        printf("Pilha vazia\n");
    }
    int resp = p->array[p->topo--];
    return resp;
}

void mostrar(Pilha *p) {
    if(p->topo == -1) {
        printf("Pilha vazia\n");
    }
    for(int i = p->topo; i >= 0; i--) {
        printf("%d ", p->array[i]);
    }
    printf("\n");
}
void freePilha(Pilha *p) {
    free(p->array);
}

int main() {
    Pilha pilha;
    criarPilha(&pilha, 5);
    inserir(&pilha, 7);
    inserir(&pilha, 3);
    inserir(&pilha, 9);
    mostrar(&pilha);
    remover(&pilha);
    mostrar(&pilha);
    inserir(&pilha, 10);
    mostrar(&pilha);
    freePilha(&pilha);
    return 0;
}

File: segundoPeriodo/AEDS-2/estudos/tadsLineares/pilha/java/Main.java
import java.util.Scanner;

class Pilha {
    private int array[];
    private int topo;
    
    public Pilha(int tamanho) {
        this.topo = -1;
        this.array = new int[tamanho];
    }
    
    public void inserir(int x) {
        if(topo == array.length - 1) {
            throw new IllegalArgumentException("Pilha cheia");
        }
        array[++topo] = x;
    }

    public int remover() {
        if(topo == -1) {
            throw new IllegalArgumentException("Pilha vazia");
        }
        int resp = array[topo--];
        return resp; 
    }

    public void mostrar() {
        if(topo == -1) {
            throw new IllegalArgumentException("Pilha vazia");
        }
        for(int i = topo; i >= 0; i--) {
            System.out.printf("%d ", array[i]);
        }
        System.out.printf("\n");
    }
}

public class Main {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        Pilha pilha = new Pilha(5);
        pilha.inserir(7);
        pilha.inserir(3);
        pilha.inserir(9);
        pilha.mostrar();
        pilha.remover();
        pilha.mostrar();
        pilha.inserir(10);
        pilha.mostrar();
    }
}